src/core/elf.d

```diff
@@ -0,0 +1,255 @@
+/**
+ * This code reads ELF files and sections using memory mapped IO.
+ *
+ * Reference: http://www.dwarfstd.org/
+ *
+ * Copyright: Copyright Digital Mars 2015 - 2015.
+ * License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   Yazan Dabain
+ * Source: $(DRUNTIMESRC core/elf.d)
+ */
+
+module core.elf;
+
+version (linux) version = linux_or_bsd;
+else version (FreeBSD) version = linux_or_bsd;
+else version (DragonFlyBSD) version = linux_or_bsd;
+
+version (linux_or_bsd):
+
+import core.sys.posix.fcntl;
+import core.sys.posix.unistd;
+
+version (linux) import core.sys.linux.elf;
+version (FreeBSD) import core.sys.freebsd.sys.elf;
+version (DragonFlyBSD) import core.sys.dragonflybsd.sys.elf;
+
+struct ElfFile
+{
+    static bool open(const(char)* path, out ElfFile file) @nogc nothrow
+    {
+        file.fd = .open(path, O_RDONLY);
+        if (file.fd < 0)
+            return false;
+
+        // memory map header
+        file.ehdr = MMapRegion!Elf_Ehdr(file.fd, 0, Elf_Ehdr.sizeof);
+        return isValidElfHeader(*file.ehdr);
+    }
+
+    @disable this(this);
+
+    ~this() @nogc nothrow
+    {
+        if (fd != -1) close(fd);
+    }
+
+    int fd = -1;
+    MMapRegion!Elf_Ehdr ehdr;
```

Comment:
Do we want these two to be directly exposed?

Reply:
`fd` perhaps not, but `ehdr` is essential.